### PR TITLE
HHH-14112 - Always join all tables from superclass entity persisters that have a WHERE condition

### PR DESCRIPTION
**Pull request by beikov**
https://hibernate.atlassian.net/browse/HHH-14112

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java [hunk 1]
```diff
@@ -136,12 +136,20 @@ public static JoinType toHibernateJoinType(int astJoinType) {
...
 		collectReferencedTables( new ASTIterator( query ), result );
-		for (FromElement fromElement : (List<FromElement>) query.getFromClause().getFromElements()) {
+		for ( FromElement fromElement : (List<FromElement>) query.getFromClause().getFromElements() ) {
 			// For joins, we want to add the table where the association key is mapped as well as that could be a supertype that we need to join
 			String role = fromElement.getRole();
 			if ( role != null ) {
 				result.add( fromElement.getOrigin().getPropertyTableName(role.substring(role.lastIndexOf('.') + 1)) );
 			}
+			EntityPersister entityPersister = fromElement.getEntityPersister();
+			if ( entityPersister != null && entityPersister instanceof AbstractEntityPersister ) {
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:146]: the nullness checking is redundant in face of the usage of 'instanceof'

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java [hunk 2]
```diff
@@ -136,12 +136,20 @@ public static JoinType toHibernateJoinType(int astJoinType) {
...
+				AbstractEntityPersister aep = (AbstractEntityPersister) entityPersister;
+				while ( !aep.filterFragment("", Collections.emptyMap()).isEmpty() && aep.getMappedSuperclass() != null ) {
+					Collections.addAll(result, aep.getTableNames());
+					aep = (AbstractEntityPersister) walker.getSessionFactoryHelper().findEntityPersisterByName( aep.getMappedSuperclass() );
+				}
+			}
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:152]: seems we can add quite some space chars to align with existing code format above

**Comment by beikov** [comment]
Hey @NathanQingyangXu 
I just quickly did this fix in the Web UI to see if tests would fail, but I'm kinda busy. Would be awesome if you could take over from here on :)

**Comment by NathanQingyangXu** [comment]
Thanks. I will try. Do I have rights to push to the git branch then?

**Comment by beikov** [comment]
I'm not sure, but it seems I pushed the branches to the wrong repo anyway, so you can create a new branch on your fork and create a new PR. Then we can close this and remove the branch from the main repo

**Comment by NathanQingyangXu** [comment]
sure. my only concern is if testing failed, I might bother you to take a
look.

**Comment by beikov** [comment]
>sure. my only concern is if testing failed, I might bother you to take a look.

No problem, I will help, but as you saw already via Travis, all tests are green.

**Comment by NathanQingyangXu** [comment]
> > sure. my only concern is if testing failed, I might bother you to take a look.
> 
> No problem, I will help, but as you saw already via Travis, all tests are green.

I meant the new testing case provided by bug reporter. Anyway, I'll try. But as I am free now, it might take some time to finalize the new PR.

**Comment by NathanQingyangXu** [comment]
new PR created at: https://github.com/hibernate/hibernate-orm/pull/3481

**Comment by beikov** [comment]
Superseded by https://github.com/hibernate/hibernate-orm/pull/3481

**Pull request closed without merging**